repose-aggregator/commons/utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletResponseWrapper.scala

```diff
@@ -63,67 +76,275 @@ class HttpServletResponseWrapper(originalResponse: HttpServletResponse, headerMo
    * @param bodyMode the mode to use for body accessibility
    */
   def this(originalResponse: HttpServletResponse, headerMode: ResponseMode, bodyMode: ResponseMode) =
     this(originalResponse, headerMode, bodyMode, originalResponse.getOutputStream)
 
-  override def getHeaderNamesList: util.List[String] = ???
+  object ResponseBodyType extends Enumeration {
+    val Available, OutputStream, PrintWriter = Value
+  }
 
-  override def addHeader(headerName: String, headerValue: String, quality: Double): Unit = ???
+  private val bodyOutputStream = bodyMode match {
+    case ResponseMode.PASSTHROUGH => new PassthroughServletOutputStream(desiredOutputStream)
+    case ResponseMode.READONLY => new ReadOnlyServletOutputStream(desiredOutputStream)
+    case ResponseMode.MUTABLE => new MutableServletOutputStream(desiredOutputStream)
+  }
 
-  /**
-   * @throws IllegalStateException when headerMode is anything other than ResponseMode.MUTABLE
-   */
-  override def appendHeader(headerName: String, headerValue: String): Unit = ???
+  private val caseInsensitiveOrdering = Ordering.by[String, String](_.toLowerCase)
+  private lazy val bodyPrintWriter = new PrintWriter(bodyOutputStream)
+
+  private var headerMap: Map[String, Seq[String]] = new TreeMap[String, Seq[String]]()(caseInsensitiveOrdering)
+  private var responseBodyType = ResponseBodyType.Available
+
+  override def getResponse: ServletResponse = throw new UnsupportedOperationException("getResponse is not supported")
+
+  override def setResponse(servletResponse: ServletResponse): Unit =
+    throw new UnsupportedOperationException("setResponse is not supported")
+
+  override def getHeaderNamesList: util.List[String] = getHeaderNames.toList
+
+  override def getHeaderNames: util.Collection[String] = headerMap.keys
+
+  override def containsHeader(name: String): Boolean = headerMap.contains(name)
+
+  override def getHeader(name: String): String = getHeaderValues(name).headOption.orNull
+
+  override def getHeaders(name: String): util.Collection[String] = getHeaderValues(name)
+
+  override def getHeadersList(name: String): util.List[String] = getHeaderValues(name)
+
+  override def getPreferredHeaders(name: String): util.List[String] = getPreferredHeaderValues(getHeaderValues(name))
+
+  override def getPreferredSplittableHeaders(name: String): util.List[String] =
+    getPreferredHeaderValues(getSplittableHeaders(name))
+
+  override def getSplittableHeaders(name: String): util.List[String] =
+    getHeaderValues(name).foldLeft(List.empty[String])((list, value) => list ++ value.split(","))
+
+  override def addHeader(name: String, value: String): Unit = {
+    headerMap = headerMap + (name -> (headerMap.getOrElse(name, Seq.empty[String]) :+ value))
+
+    if (headerMode != ResponseMode.MUTABLE) {
+      // Write through to the wrapped response immediately
+      super.addHeader(name, value)
+    }
+  }
+
+  override def addHeader(name: String, value: String, quality: Double): Unit =
```

Comment:
I guess I'm leaning toward yes now, but you tell me.

I will say that the q value according to HTTP spec only allows up to three decimal points. That limits the quality domain to 10^3 + 1 distinct values. Not range checking obviously expands that domain, but honestly, there should never be a need.

On the usage side, do we care what a user stick heres? Our code always just takes the highest value. If we do range check, do we through a compile time or runtime exception? In the former case, do I want to deal with that every time I call this code? Probably not. In the latter case, if I haven't tested my code very well I start getting exceptions at runtime. Is an invalid quality an unrecoverable exception? The HTTP spec does define a quality value as being in the range of 0 to 1, but it does not preclude the existence of a parameter named 'q' being used to convey any value in the general case.

For reference:
http://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html#sec3.9